ocaml/typing/typeopt.ml

```diff
@@ -196,39 +198,79 @@ let value_kind_of_value_layout layout =
   | Immediate -> Pintval
   | Immediate64 ->
     if !Clflags.native_code && Sys.word_size = 64 then Pintval else Pgenval
-  | Any | Void -> assert false
+  | Any | Void | Float64 -> assert false
 
-(* Invariant: [value_kind] functions may only be called on types with layout
-   value. *)
-(* CR layouts v2: However, the current version allows for it to be called on
-   types whose layouts we can not determine due to missing cmis.  This is
-   unsound once we have non-value layouts.
+(* [value_kind] tries to maintain the invariant that it is only called on
+   values.  With the current set of layout restrictions, there are two reasons
+   this invariant may be violated:
 
-   For the particularly annoying case where we're missing the cmis for
-   subcomponent of a composite type, we'll need to fall back and return Pgenval
-   for the whole type.  See this commit, which implements that:
+   1) A bug in the type checker or the translation to lambda.
+   2) A missing cmi file, so that we can't accurately compute the layout of
+      some type.
 
-   https://github.com/ccasin/ocaml-jst/tree/unboxed-types-v1-layouts-fallback
-*)
-(* CR layouts v5 (and maybe other versions): As we relax the requirement that
-   various things have to be values, many recursive calls in [value_kind]
-   need to be updated to check layouts. *)
-(* CR layouts v2: At the moment, sort variables may be defaulted to value by
-   value_kind.  For example, here:
+   In case 1, we have a bug and want to issue a scary error.
+
+   In case 2, we could issue an error and make the user add the dependency
+   explicitly.  But because [value_kind] looks at the subcomponents of your type,
+   this can lead to some surprising and unnecessary errors.  Suppose we're
+   computing the value kind for some type:
+
+     type t = int * M.t
+
+   If we're missing the cmi for [M], we can't verify the invariant that
+   [value_kind] is only called on values.  However, we still know the pair
+   itself is a value, so a sound thing to do is fall back and return [Pgenval]
+   for [t].
+
+   On the other hand, if we're asked to compute the value kind for [M.t]
+   directly and are missing the cmi for [M], we really do need to issue an error.
+   This is a bug in the typechecker, which should have checked that the type
+   in question has layout value.
+
+   The implementation of this fallback behavior is combined with the layouts
```

Comment:
@ccasin and I discussed a restructuring of this comment, based around the fact that the glorious future will require layout checks in advance of calling `value_kind` -- but these aren't "safety" or "pre-condition" checks as much as just a control flow that will sometimes call `value_kind` and sometimes won't.

Reply:
I've taken a stab at updating these - see what you think.

As an aside, I'm starting to think we should plan to split the "unboxed types in structures" work into two parts: first doing unboxed types in declared structures like records and variants, where layout information is available in the declaration, and following up with things like tuples and polymorphic variants that will require either changes to carry layouts in types or unfortunate layout checking in transl.